### DISCUSSION TITLE
CustomEvents in addEventListener

### DISCUSSION THREAD

**Discussion by grzegorzkiselev** [voxpelli/types-in-js#24] [Questions]
Hey, everybody. I am trying to understand the typing of custom events.

<img width="601" alt="Снимок экрана 2023-12-21 в 14 36 25" src="https://github.com/voxpelli/types-in-js/assets/61684506/9e65b50c-7cb5-4032-8046-6b1ccc9c4b9e">
<img width="752" alt="Снимок экрана 2023-12-21 в 14 36 34" src="https://github.com/voxpelli/types-in-js/assets/61684506/90e01434-5ad6-4e44-9c7b-556e53318103">

I encountered that I can’t specify “CustomEvent” type in addEventListener without patching TS library. Any ideas what I’m doing wrong?

<img width="941" alt="Снимок экрана 2023-12-21 в 14 37 22" src="https://github.com/voxpelli/types-in-js/assets/61684506/aa32589f-a009-4eb3-954b-be15ac6e4f75">

The next question goes further: what can I get by specifying `@fires`, `@listens` tags and declaring events via `@event`? Does this give me any hints, or just display information in the popup?

**Comment by samthor** [comment]
I don't know if this is what you want, but I wrote up a way in TS (yes, I know this is "types-in-JS", but does importing .d.ts files count?) to somewhat neatly add new event types to classes:

https://github.com/samthor/typed-event-types

**Reply by grzegorzkiselev** [reply]
Thanks for the answer! Your result was more complicated than I expected. What's confusing to me is that Typescript is not prepared to accept a CustomEvent as an EventListener's callback parameter. Are there any good reasons why it doesn't?
